### DISCUSSION TITLE
Who wants a scrollbar in VisiData?

### DISCUSSION THREAD

**Discussion by saulpw** [saulpw/visidata#1140] [Ideas]
Talking with @jsvine this past weekend made me realize how spatially-oriented some people are.  I myself don't have a strong spatial sense, I apparently only think in terms of the extremes (like top/bottom/leftmost/rightmost), but of course I'd like VisiData to accommodate a broad range of styles if enough people are interested.  So, let's try this out: if this post gets at least 10 upvotes/likes, I'll bang out a right-side visual scrollbar to indicate cursor position in the rows.  And I'll add mouse click support if it gets at least 20 upvotes.

**Comment by ghost** [comment]
It's not exactly the same thing, but I was thinking recently I'd love a percentage indicator to tell me how far through a sheet I am.
Something like this from vim:

![image](https://user-images.githubusercontent.com/55477181/136898307-ccdfc40c-db3a-49e7-a712-3453ff22b7f6.png)

I suspect I'd want to disable a scrollbar to maximise screen real-estate!

**Reply by saulpw** [reply]
Put this in your .visidatarc:

```
@Sheet.property
def pctRows(sheet):
    return sheet.cursorRowIndex*100//(sheet.nRows or 1)

Sheet.class_options.disp_rstatus_fmt+='  {sheet.pctRows:02d}%'
```

I think the scrollbar should only take up one character on the right-side.  Not a lot, but might be a lot of real-estate in some environments, so of course it would be disable-able.

**Reply by ghost** [reply]
Amazing! Thank you. :)

I tweaked it a bit as cursorRowIndex starts at 0, moved the percentage conversion to the format string, and changed the formatting string so it always takes up the same amount of space:

``` python

# Progress indicator
@Sheet.property
def pctRows(sheet):
    return (sheet.cursorRowIndex + 1)/(sheet.nRows or 1)

Sheet.class_options.disp_rstatus_fmt += '  {sheet.pctRows:>4.0%}'
```

**Reply by saulpw** [reply]
I updated our examples to use `Sheet.class_options` so that there's not a dangling `%` on the status bar for non-table Sheets (like the Graph).

**Comment by frosencrantz** [comment]
I trust Visidata to be my camera on data.  I expect it to be accurate and truthful.  Data doesn’t always fit in the screen space provided, so it is important for Visidata to be able to make it clear when there is more data than shown. Overall this is done really well, for example ellipses when a value expands beyond the edge of a cell, or the visual markers when there are more columns.  And for structured data like json where a cell can be an array or an object, it is visually clear that is more data to explore. 

There was one case  where Visidata misdirected me about where data was hidden, when the sqlite loader would to report that an unloaded table had 0 rows and 1 column (https://github.com/saulpw/visidata/issues/479).  This misdirection caused me to not try to explore this data. This was fixed, but there is a recent regression for some tables and views (https://github.com/saulpw/visidata/issues/1111) 

A scroll bar would be more helpful to show that there are more rows either above or below the current view.  Though I would be happy with a solution just to make it clear about the vertical edges of the data, and where it continues off the edges.

**Reply by saulpw** [reply]
This is good feedback, @frosencrantz.  Let's see if we can get to 10 upvotes!

**Comment by geekscrapy** [comment]
I'd love the scroll bar, and I'm quite surprised this hasn't already got more votes! Although, I will say the % indicator that you all came up with is a nice compromise.

I think it'd be great to have the % included by default if the scrollbar isn't built

**Comment by geekscrapy** [comment]
In the same sort of vein, and I appreciate this a moonshot, but I always find the mini maps you find in some IDEs and editing apps extremely useful as you can see the shape of the data. I have absolutely no idea how to even represent this though 😂

**Reply by saulpw** [reply]
Maybe something using braille chars?  Would you want to see values vs nulls, or what?

**Reply by geekscrapy** [reply]
Braille characters would be good. I use it in other apps to see the length of the lines, which would be extremely odd to do within vd. The length of the lines show me (generally!) changes in the pattern of the data (e.g. heartbeat requests of an application within web traffic logs etc) and I can then immediately zoom in at points where the rhythm/length deviates, which usually indicates something different happening at that time

**Reply by saulpw** [reply]
Can you use `z#` to type the text as `len` (or `=len(text)`), and then `.` to graph?  How many rows are in the set?

**Reply by geekscrapy** [reply]
That's a good idea!
The number of rows varies wildly, sometimes millions, sometimes 1000s. I often use the minimap and the first few lines to gauge if a file is even worth looking at, just from this pattern straight after opening it. If in doubt, I'll move down through the lines to make sure my assumptions are correct. For this reason I would kind of need to be able to see the minimap graph and the original logs at the same time. Maybe this is something that can be achieved with the split pane?!

**Reply by saulpw** [reply]
Seems possible :)

**Reply by geekscrapy** [reply]
An example as I have it open - this is a minimap from vscode. Left side is the end of the logs, and the minimap is on the right. You can see the general flow of the taffic is extremely similar apart from 4 distinct points. This narrows down analysis efforts considerably.

Something I was trying to dev was a `go-to-next-diff` command which would help. The idea was that you would define a % tolerance, and use the `go-to-next` logic, but instead of "is it equal", "is the value % different to the current". Maybe I should resurrect this...

![image](https://user-images.githubusercontent.com/11225502/140020871-d5d08357-92fa-4b11-b04f-d8085f4b4822.png)

**Reply by ghost** [reply]
It might be interesting to have a few different ways of representing zoomed out rows. e.g.

- NULL/NOT NULL for each column with blank or `.` (using braille characters for high resolution): `⠪⠓⠭`
- Count the total number of characters in the row and make a histogram
- Count the total number of NOT NULL columns in the row and make a histogram
- Segment the data structure into 'squares' of n columns X n rows. For each segment, count one of the above measures, and display either a `.` or a blank depending on a threshold.
